Selaraskan assert state usage promosi ke done di uji enrollment

## Skill yang Wajib Dimuat

Muat skill berikut lewat `Skill` tool **sebelum** mengerjakan langkah mana pun:

- `odoo-development-unit-test` — merevisi skenario `odoo-yaml-test` yang gagal
- `odoo-development-repo-ops` — commit per-modul, PR ke branch `14.0`, merge `/ocabot`

## Konteks

Skenario `Approving the usage reduces the term's own invoice residual` di `tests/test_data_school_enrollment_payment_term_promotion.yaml` gagal di CI **tanpa satu baris pun kode ssi-school berubah**, dengan `YamlAssertionError: promotion_code_usage.state: expected 'open', got 'done'` pada step `Approve the usage` (assert di baris 747). Penyebabnya kontrak modul hulu: `ssi_promotion` naik `14.0.1.9.1` → `14.0.1.17.0`, dan commit `b26be31` (`Closes open-synergy/ssi-promotion#72`) menambahkan override `action_open()` pada `ssi_promotion/models/promotion_code_usage.py` yang memanggil `action_done()` sendiri begitu `recognition_state` bernilai `not_applicable` — yaitu saat `amount_to_recognize` nol karena tak ada sisi Deferred. Perubahan itu **disengaja dan terdokumentasi**: dokumen Instruksi Kerja modul hulu untuk aksi Approve pada `promotion_code_usage` (berkas `05-approve.md` di repo `ssi-promotion`, bukan di repo ini) menetapkan dua cabang Post-Condition (tanpa sisi deferred → langsung Done; ada sisi deferred → berhenti di Open), hulu mencabut tombol Done manual (`_automatically_insert_done_button` bernilai `False`), dan menyertakan migrasi `ssi_promotion/migrations/14.0.1.17.0/post-migrate.py` yang men-stamp baris lama `open` tanpa sisa deferred menjadi `done`. Test hulu sendiri sudah diselaraskan (26 blok assert `done`; yang masih `open` khusus skenario Deferred). Bukti drift: eksperimen kontrol run `31980705647` di base `14.0` SHA `ef3c64c` — attempt 1 hijau, attempt 2 merah, dan diff `pip freeze` atas 150 paket hanya menyisakan 2 baris berbeda.

## Sasaran

Skenario itu kembali hijau dengan maksud aslinya utuh: menyetujui usage promosi mengurangi residual journal item receivable milik payment term-nya, dan state akhir usage-nya mencerminkan kontrak hulu yang baru untuk promotion type Immediate.

## Ruang Lingkup

- **Modul (Odoo):** `ssi_school_promotion`
- **Repo (owner/repo):** `open-synergy/ssi-school`
- **Versi Odoo (branch target):** `14.0`
- **Model:** `promotion_code_usage`, `school_enrollment_payment_term`

**Permukaan perubahan:** `tests/` — hanya berkas skenario YAML; tak ada kode modul yang berubah

**Artefak terdampak:** `tests/test_data_school_enrollment_payment_term_promotion.yaml`, skenario `Approving the usage reduces the term's own invoice residual` (1 skenario, 1 nilai assert) — **ikut di item ini**

**Takaran:** uji=1

```
#BUDGET model=sonnet window=1000000 ambang=225000 ceiling=850000 reserve=26000 protocol=10614 skill=40527 repo=8283 work=9800 total=95224 pct=42 class=S verdict=FIT warns=1 modul=ssi_school_promotion versi=14.0
```

**Termasuk:**

- Mengubah nilai harapan assert `state` pada step `Approve the usage` di skenario itu dari `open` menjadi `done`.

**Tidak termasuk:**

- Mengubah fixture `promotion_type` skenario itu menjadi Deferred — **dilarang**; lihat `## Keputusan Desain`.
- Empat skenario lain di berkas itu — tak satu pun menyetujui sebuah `promotion_code_usage`.
- Lima belas assert `state` bernilai `open` lain di berkas itu — semuanya milik model lain (`school_enrollment`, `customer_invoice`, `promotion_code`) dan **tetap** `open`.
- Modul kembarannya untuk admission — dicatat sebagai item terpisah (lihat `## Dependensi`).
- Repo hulu `ssi-promotion`: dilarang disentuh, perubahannya disengaja.
- Menghapus, men-skip, atau melonggarkan assertion mana pun demi menghijaukan CI.

## Keputusan Desain

- **Hanya nilai assert `state` yang berubah** — dari `open` menjadi `done` — pada satu step saja: `Approve the usage` di skenario `Approving the usage reduces the term's own invoice residual`. Tak ada step yang ditambah, dihapus, atau diurutkan ulang; tak ada fixture yang disentuh; tak ada nama step yang perlu berubah.
- **Fixture skenario itu memang Immediate, dan itulah alasan `done` benar.** Step `Create promotion type allowing school_enrollment_payment_term as reference` membuat `promotion_type` tanpa menyetel `recognition_method` maupun `referrer_recognition_method`, sehingga keduanya jatuh ke default hulu — `immediate` dan `same`. Akibatnya `_get_amount_to_recognize_raw()` mengembalikan `0.0`, `recognition_state` menjadi `not_applicable`, dan override `action_open()` hulu memanggil `action_done()` di dalam transisi yang sama. `done` adalah state yang benar menurut kontrak baru, bukan penyesuaian demi menghijaukan CI.
- **Mengubah fixture menjadi Deferred DILARANG.** Skenario ini menguji pengurangan residual pada promosi Immediate; menggesernya ke Deferred akan mengganti subjek uji, bukan memperbaikinya.
- **Transisi ke `done` tidak menggeser assert lain di skenario ini, dan assert berikutnya tidak boleh diubah.** Assert penutup skenario — `move_line_a.amount_residual` bernilai `800000.0` — adalah inti skenarionya dan tetap benar: seluruh efek samping (pembuatan & posting accounting entry, rekonsiliasi alokasi, penomoran sequence) dijalankan `super().action_open()` **sebelum** `action_done()` dipanggil. `promotion_code_usage` hulu juga tidak memasang satu pun hook `pre_done_check`/`post_done_action`, jadi `action_done()` hanya membalik state; tak ada field yang baru terisi saat Done yang perlu di-assert. `ssi_school_promotion` sendiri tidak meng-inherit `promotion_code_usage` sama sekali.
- **Assert `state` bernilai `open` lain di berkas ini tetap apa adanya.** Di berkas ini ada 16 assert `state` bernilai `open`; hanya satu (baris 747) yang menyasar `promotion_code_usage`. Lima belas sisanya menyasar `school_enrollment`, `customer_invoice`, dan `promotion_code` — model yang kontraknya tidak berubah.

## Rencana Eksekusi

1. [odoo-development-unit-test] Ubah nilai harapan assert `state` pada step `Approve the usage` di skenario `Approving the usage reduces the term's own invoice residual` menjadi `done`.
2. [odoo-development-unit-test] Jalankan `validate-unit-test.sh` dan `verify-tests.sh` pada modul ini sampai keduanya bersih.
3. [odoo-development-repo-ops] Commit per-modul + PR ke branch `14.0`; pantau CI; merge via `/ocabot merge`.

## Kriteria Penerimaan

- [ ] Step `Approve the usage` di skenario itu meng-assert `state` bernilai `done`
- [ ] Fixture `promotion_type` skenario itu tetap tanpa `recognition_method` dan tanpa `referrer_recognition_method` (tetap Immediate)
- [ ] Assert penutup skenario — `move_line_a.amount_residual` bernilai `800000.0` — tidak berubah satu karakter pun
- [ ] Berkas itu masih memuat tepat 15 assert `state` bernilai `open`, tak satu pun menyasar `promotion_code_usage`
- [ ] Empat skenario lain di berkas itu tidak berubah satu baris pun
- [ ] Tidak ada step yang ditambahkan, dihapus, atau diurutkan ulang di berkas itu
- [ ] Job `tests` di CI hijau untuk modul ini

## Skenario Uji

- Positif: sesudah `action_approve_approval`, usage promosi bertipe Immediate atas `school_enrollment_payment_term` berakhir di state `done` — dilewatkan `open` oleh override `action_open()` hulu — dan residual journal item receivable milik term-nya sudah berkurang sebesar diskon, dari `1000000.0` menjadi `800000.0`.
- Negatif: assert residual `800000.0` adalah penjaga sisi negatifnya — bila accounting entry diskon gagal terbentuk atau gagal direkonsiliasi ke faktur term, residual tetap `1000000.0` dan skenario gagal alih-alih lolos diam-diam meski state-nya sudah `done`.

## Definition of Done

- [ ] Seluruh Kriteria Penerimaan tercentang
- [ ] Docstring lengkap di **setiap** class & method yang ditulis/diubah — termasuk file `tests/` — bahasa Inggris, gaya reST, ≤ 79 karakter (skill `odoo-development`, `10-docstring.md`); diverifikasi dengan pemeriksa AST di `12-checklists.md`, bukan dari ingatan
- [ ] Unit test ditulis (skenario `odoo-yaml-test`), dijalankan di CI GitHub (bukan lokal)
- [ ] Tidak ada test/tour/assertion yang dihapus, di-skip, atau dilonggarkan demi menghijaukan CI
- [ ] Modul ter-install/update tanpa error (hanya bila user minta uji lokal)
- [ ] Commit per-modul + PR ke branch `14.0`; CI hijau; merge via `/ocabot merge`
- [ ] Issue ditutup dengan reason `completed`
- [ ] Pasangan lintas-versi (bila ada) sudah diberi tahu lewat komentar

## Dependensi

Tidak menunggu issue lain — `## Rencana Eksekusi` dapat dijalankan apa adanya hari ini.

Kembarannya untuk modul admission: `open-synergy/ssi-school#348` — bukan prasyarat, sebab keputusannya ditulis lengkap di kedua issue. Keduanya berada di repo dan versi yang sama, jadi pendaratannya diserialkan orkestrator, bukan oleh `Depends on:`.

**Memblokir:** bersama `open-synergy/ssi-school#345` dan `open-synergy/ssi-school#346`, item ini memblokir seluruh backlog audit ssi-school 14.0. Selama suite repo merah, PR audit mana pun tertahan — mis. `open-synergy/ssi-school#336` yang PR-nya `open-synergy/ssi-school#344` menunggu karena kegagalan ini, bukan karena cacat di modulnya sendiri.

## Tautan Lintas Versi

—

--- a/ssi_school_admission_customer_invoice_export/tests/test_data_admission_create_invoice_export.yaml
+++ b/ssi_school_admission_customer_invoice_export/tests/test_data_admission_create_invoice_export.yaml
@@ -352,13 +352,90 @@ scenarios:
             type: "value"
             expected: "open"
 
-      - step: "Mark invoice Done as fully paid"
+      - step: "Setup - create contra account for the manual payment entry"
+        action: "create"
+        model: "account.account"
+        save_as: "payment_contra_account"
+        values:
+          name: "Payment Contra Account AI1"
+          code: "PCAAI1"
+          user_type_id: "REC: account_type.id"
+
+      - step: "Setup - create general journal for the manual payment entry"
+        action: "create"
+        model: "account.journal"
+        save_as: "payment_journal"
+        values:
+          name: "Payment General Journal AI1"
+          code: "PGJAI1"
+          type: "general"
+
+      - step: "Get invoice Done's receivable journal item"
+        action: "search"
+        model: "account.move.line"
+        domain: [["id", "=", "REC: invoice_done.receivable_move_line_id.id"]]
+        save_as: "invoice_done_receivable_ml"
+        expect_count: 1
+
+      - step: "Create a manual payment entry crediting the receivable account"
+        action: "create"
+        model: "account.move"
+        save_as: "payment_move"
+        values:
+          journal_id: "REC: payment_journal.id"
+          date: "2024-07-01"
+          line_ids:
+            - - 0
+              - 0
+              - name: "Payment"
+                account_id: "REC: ci_account.id"
+                partner_id: "REC: invoice_done.partner_id.id"
+                credit: 500000.0
+            - - 0
+              - 0
+              - name: "Payment contra"
+                account_id: "REC: payment_contra_account.id"
+                debit: 500000.0
+
+      - step: "Post the payment entry"
         action: "call"
+        target: "payment_move"
+        method: "action_post"
+
+      - step: "Get the payment's receivable-side journal item"
+        action: "search"
+        model: "account.move.line"
+        domain:
+          - ["move_id", "=", "REC: payment_move.id"]
+          - ["account_id", "=", "REC: ci_account.id"]
+        save_as: "payment_receivable_ml"
+        expect_count: 1
+
+      - step: "Combine both journal items to reconcile"
+        action: "search"
+        model: "account.move.line"
+        domain:
+          - [
+              "id",
+              "in",
+              ["REC: invoice_done_receivable_ml.id", "REC: payment_receivable_ml.id"],
+            ]
+        save_as: "to_reconcile"
+        expect_count: 2
+
+      - step: "Reconcile the receivable line against the payment"
+        action: "call"
+        target: "to_reconcile"
+        method: "reconcile"
+
+      - step: "Invoice Done becomes done once fully reconciled"
+        action: "assert"
         target: "invoice_done"
-        method: "action_done"
-        as_user: "base.user_admin"
         refresh: true
         asserts:
+          realized:
+            type: "value"
+            expected: true
           state:
             type: "value"
             expected: "done"

--- a/ssi_school_admission_promotion/tests/test_data_school_admission_payment_term_promotion.yaml
+++ b/ssi_school_admission_promotion/tests/test_data_school_admission_payment_term_promotion.yaml
@@ -705,7 +705,7 @@ scenarios:
         asserts:
           state:
             type: "value"
-            expected: "open"
+            expected: "done"
       - step: "The term's own receivable residual is reduced by the discount"
         action: "assert"
         target: "move_line_a"

--- a/ssi_school_customer_invoice_export/tests/test_data_enrollment_create_invoice_export.yaml
+++ b/ssi_school_customer_invoice_export/tests/test_data_enrollment_create_invoice_export.yaml
@@ -389,13 +389,90 @@ scenarios:
             type: "value"
             expected: "open"
 
-      - step: "Mark invoice Done as fully paid"
+      - step: "Setup - create contra account for the manual payment entry"
+        action: "create"
+        model: "account.account"
+        save_as: "payment_contra_account"
+        values:
+          name: "Payment Contra Account IE1"
+          code: "PCAIE1"
+          user_type_id: "REC: account_type.id"
+
+      - step: "Setup - create general journal for the manual payment entry"
+        action: "create"
+        model: "account.journal"
+        save_as: "payment_journal"
+        values:
+          name: "Payment General Journal IE1"
+          code: "PGJIE1"
+          type: "general"
+
+      - step: "Get invoice Done's receivable journal item"
+        action: "search"
+        model: "account.move.line"
+        domain: [["id", "=", "REC: invoice_done.receivable_move_line_id.id"]]
+        save_as: "invoice_done_receivable_ml"
+        expect_count: 1
+
+      - step: "Create a manual payment entry crediting the receivable account"
+        action: "create"
+        model: "account.move"
+        save_as: "payment_move"
+        values:
+          journal_id: "REC: payment_journal.id"
+          date: "2024-07-01"
+          line_ids:
+            - - 0
+              - 0
+              - name: "Payment"
+                account_id: "REC: ci_account.id"
+                partner_id: "REC: invoice_done.partner_id.id"
+                credit: 500000.0
+            - - 0
+              - 0
+              - name: "Payment contra"
+                account_id: "REC: payment_contra_account.id"
+                debit: 500000.0
+
+      - step: "Post the payment entry"
         action: "call"
+        target: "payment_move"
+        method: "action_post"
+
+      - step: "Get the payment's receivable-side journal item"
+        action: "search"
+        model: "account.move.line"
+        domain:
+          - ["move_id", "=", "REC: payment_move.id"]
+          - ["account_id", "=", "REC: ci_account.id"]
+        save_as: "payment_receivable_ml"
+        expect_count: 1
+
+      - step: "Combine both journal items to reconcile"
+        action: "search"
+        model: "account.move.line"
+        domain:
+          - [
+              "id",
+              "in",
+              ["REC: invoice_done_receivable_ml.id", "REC: payment_receivable_ml.id"],
+            ]
+        save_as: "to_reconcile"
+        expect_count: 2
+
+      - step: "Reconcile the receivable line against the payment"
+        action: "call"
+        target: "to_reconcile"
+        method: "reconcile"
+
+      - step: "Invoice Done becomes done once fully reconciled"
+        action: "assert"
         target: "invoice_done"
-        method: "action_done"
-        as_user: "base.user_admin"
         refresh: true
         asserts:
+          realized:
+            type: "value"
+            expected: true
           state:
             type: "value"
             expected: "done"

--- a/ssi_school_promotion/tests/test_data_school_enrollment_payment_term_promotion.yaml
+++ b/ssi_school_promotion/tests/test_data_school_enrollment_payment_term_promotion.yaml
@@ -744,7 +744,7 @@ scenarios:
         asserts:
           state:
             type: "value"
-            expected: "open"
+            expected: "done"
       - step: "The term's own receivable residual is reduced by the discount"
         action: "assert"
         target: "move_line_a"

